libnavigation-core/src/test/java/com/mapbox/navigation/core/MapboxNavigationTest.kt

```diff
@@ -1180,19 +1183,42 @@ class MapboxNavigationTest {
     }
 
     @Test
-    fun `set route - immediately stops the refresh controller`() = coroutineRule.runBlockingTest {
+    fun `route refresh of previous route completes after new route is set`() = coroutineRule.runBlockingTest {
+        every { NavigationComponentProvider.createDirectionsSession(any()) } answers {
+            MapboxDirectionsSession(mockk(relaxed = true))
+        }
         createMapboxNavigation()
-
-        val shortRoutes = listOf<NavigationRoute>(mockk())
-        coEvery { tripSession.setRoutes(shortRoutes, any(), any()) } coAnswers {
-            delay(50L)
+        val first = listOf(createNavigationRoute(createDirectionsRoute(requestUuid = "test1")))
+        val second = listOf(createNavigationRoute(createDirectionsRoute(requestUuid = "test2")))
+        val refreshOrFirstRoute = CompletableDeferred<Unit>()
+        coEvery { routeRefreshController.refresh(any()) } coAnswers {
+            CompletableDeferred<Unit>().await() // never completes
+            firstArg()
+        }
+        coEvery { routeRefreshController.refresh(first) } coAnswers {
+            refreshOrFirstRoute.await()
+            listOf(createNavigationRoute(createDirectionsRoute(requestUuid = "test1.1")))
+        }
+        coEvery { tripSession.setRoutes(second, any(), any()) } coAnswers {
             mockk(relaxed = true)
         }
 
-        pauseDispatcher {
-            mapboxNavigation.setNavigationRoutes(shortRoutes)
-            verify(exactly = 1) { routeRefreshController.stop() }
+        val routesUpdates = mutableListOf<RoutesUpdatedResult>()
+        mapboxNavigation.registerRoutesObserver {
+            routesUpdates.add(it)
         }
+        mapboxNavigation.setNavigationRoutes(first)
+        mapboxNavigation.setNavigationRoutes(second)
+        refreshOrFirstRoute.complete(Unit)
```

Comment:
Isn't this cheating the test?

If the first refresh did not complete yet, it should be canceled when the second route is set rather than actually complete successfully.

Reply:
Doesn't seem like a cheating. If route refresh cancelled,  completion of `refreshOrFirstRoute` doesn't change anything, see a simple example https://pl.kotl.in/RZwTdZPcD